Port the tables and trees sections of the tskit tut

hyanwong opened the pull request:
Part of https://github.com/tskit-dev/tskit/issues/1121, this moves a substantial section of what was previously the ["tutorial" section](https://tskit.dev/tskit/docs/stable/tutorial.html) into https://tskit.dev/tutorials/data_structures.html (see https://github.com/tskit-dev/tutorials/issues/36).

I haven't made much of an attempt to tidy up the flow, so it's a bit of a rag-bag at the moment, although it does roughly fit into the framework that @jeromekelleher suggested of of "Tables" then "Trees". Perhaps he could look over to see what needs moving elsewhere (and add suggestions of where to move it).

I suspect we should massage this to add stuff from https://tskit.dev/tskit/docs/stable/data-model.html, but this is hopefully a good start. It's all in jupyterbook format anyway.

jeromekelleher commented on data/examples.py:
```diff
@@ -0,0 +1,457 @@
+"""
```
We shouldn't need this file - we should be doing all these code chunks in the jupyter book.

hyanwong replied:
But we don't want to actually simulate it, right? Because then it might change depending on e.g. the GSL version. So if it was created by msprime, and exact topology is important, we should use a saved `.ts` file, I assume?

jeromekelleher replied:
Yes, I think the pattern of having a hidden cell with code to generate the example ts and saving the ts in the data directory is the way to go. So we don't need this file here.

jeromekelleher commented on data_structures.md:
```diff
@@ -11,12 +11,601 @@ kernelspec:
+
+Since columns cannot be modified directly as properties of the tables,
+they must be extracted, modified, then replaced.
+:::{todo}
```
I don't think it's true anyway, you can do things like ``n.flags = sv``, can't you?

jeromekelleher replied:
Probably easier to put in an idiomatic example here now rather than update the TODO

hyanwong replied:
@benjeffery - can you work out a small bit of replacement text for this bit, as you know what the new getter and setter methods are and what the correct idiom is?

jeromekelleher replied:
I think it's literally ``table.time = time_array`` will do it. This calls set_columns under the hood

hyanwong replied:
Oh, right, I thought we might want to showcase the slicing operators

jeromekelleher replied:
OK, but that's a separate thing. I just don't want to put in code here that's telling people to do things that we actively want to discourage

jeromekelleher commented on data_structures.md:
````diff
@@ -11,12 +11,601 @@ kernelspec:
+definition of how a tree sequence is stored in a collection of tables.
+Here we give an overview. Consider the following sequence of trees:
+
+```{code-cell} ipython3
````
I don't think you need the ``ipython3``s on the code-cells

hyanwong replied:
I was copying https://jupytext.readthedocs.io/en/latest/formats.html#myst-markdown - there it says it's "optional" but useful as an aide to syntax highlighting?

jeromekelleher commented on data_structures.md:
```diff
@@ -11,12 +11,601 @@ kernelspec:
+            assert len(site.mutations) == 1  # Only supports infinite sites muts.
+            mut = site.mutations[0]
+            if tree.num_samples(mut.node) > 1:
+                site_id = tables.sites.add_row(
```
Better to update here to use the new append (I think it's called?) method. Will be a while before this is updated again, and manually specifying the cols here like this (and leaving out a bunch of them) is a major anti-pattern.

hyanwong replied:
I've done this, but doing it for mutations means using `dataclass.replace`. See comments at the end of this PR.

jeromekelleher reviewed (approved):
Good to port some of this stuff across, but I wonder if we'll be porting a lot of it back into the tskit docs when they are udpated to use the msprime-1.0 type structure.

But as you say, good to get it into JupyterBook format anyway, easy enough to port back.

We should probably remove the corresponding bits from the tskit docs, so we can keep track of what content is where (I guess putting links in to here in the relevant sections, for now?)

hyanwong commented:
> Good to port some of this stuff across, but I wonder if we'll be porting a lot of it back into the tskit docs when they are udpated to use the msprime-1.0 type structure.

Possibly - but all this stuff is in the "tutorial", and I think we probably want to remove that entire page from the tskit docs and port it into separate tutorials. As we do this, we can see what should go back into the main tskit documentation, and where.

> We should probably remove the corresponding bits from the tskit docs, so we can keep track of what content is where (I guess putting links in to here in the relevant sections, for now?)

I think first we merge these, then when we update the docs to remove them from tskit, we can change any dead refs to point to the tutorial section links instead. Since this is all in the `tutorial.html` file, which we are removing anyway, it should be easy to see what needs removing after this set of tutorial PRs are merged.

benjeffery commented:
Ping me when you've done JK's comments and I'll have a look over

hyanwong commented:
> Ping me when you've done JK's comments and I'll have a look over

I'm thinking that we might want to remove the sections on (a) iterating through trees in a ts and (b) traversing in various ways through a single tree.

(a) can probably be (mostly) incorporated into the "getting started" tut, which has a section on the `.trees()` iterator, but which doesn't mention using `reversed` (easy to stick in) or the problems of storing trees as a list (could add as a {note})`
(b) could probably go in its own tute: e.g. "tree traversal", perhaps under a new section of tuts labelled "Analysis" which could also contain the "statistical analysis" tut?

I'm not sure whether to do this now, before you look over it @benjeffery ?

benjeffery commented:
These suggestions seem good, and worth doing now. 

"The problems of storing trees as a list" can be repeated - it can be a real gotcha for newcomers.

hyanwong commented:
I hope I've addressed all the issues raised, so it's probably ready for review. The singleton deletion code uses the dataclass.replace method to set the site id: I worry a little that this is a bit opaque for a new user?

```
site_id = tables.sites.append(site)
mut = dataclasses.replace(mut, site=site_id)  # set the new site id
tables.mutations.append(mut)
```

hyanwong commented:
I kept a small section in there about "trees", as @jeromekelleher had a section titled this in the stub. But it feels a bit lonely. If it were to be removed, the entire tube is about tables & editing, so perhaps the name could be changed to, well, "Tables & editing" or something similar. I also wonder if it should go a bit lower in the list of chapters?

jeromekelleher reviewed (approved):
LGTM - I think we can merge and then start thinking again about where various bits of content should end up. Great to get the basic stuff ported over now.

benjeffery commented on getting_started.md:
```diff
@@ -104,17 +142,19 @@ plt.show()
-possibilities):
+chromosome, where the selective sweep occurred. 
+
+Currently, it's not particularly efficient to pull out individual trees from the middle
```
I'm wondering here if we should suggest people who have use cases for random access to comment in say https://github.com/tskit-dev/tskit/issues/684 so that we can know. Generally mentioning occasionally in the docs that we welcome issues/discussions seems a good thing to reduce the barriers.

hyanwong replied:
Yes, I would agree with that, as long as we note in that issue to take the comment out of the tut when the issue is closed.

hyanwong replied:
Should I do this now, before you merge it @benjeffery ?

benjeffery reviewed (approved):
Looks great @hyanwong I've been through closely and checked links.

jeromekelleher commented:
Can you rebase here please @hyanwong, we're touching a lot of files so good to be up to date with base.

hyanwong commented:
> Can you rebase here please @hyanwong, we're touching a lot of files so good to be up to date with base.

Yes, of course - rebased and force pushed. Thanks for seeing to all this. Good to get off my plate.
